checker/match.go

```diff
@@ -0,0 +1,161 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"fmt"
+	"regexp"
+
+	authz "github.com/envoyproxy/data-plane-api/api/auth"
+
+	"github.com/projectcalico/app-policy/proto"
+	"github.com/projectcalico/libcalico-go/lib/selector"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const SPIFFE_ID_PATTERN = "^spiffe://[^/]+/ns/([^/]+)/sa/([^/]+)$"
+
+var spiffeIdRegExp *regexp.Regexp
+
+// match checks if the Rule matches the request.  It returns true if the Rule matches, false otherwise.
+func match(rule *proto.Rule, req *authz.CheckRequest) bool {
+	log.Debugf("Checking rule %v on request %v", rule, req)
+	attr := req.GetAttributes()
+	return matchSource(rule, attr.GetSource()) && matchRequest(rule, attr.GetRequest())
+}
+
+func matchSource(r *proto.Rule, peer *authz.AttributeContext_Peer) bool {
+	// TODO IPSets
+	return matchServiceAccounts(r.GetSrcServiceAccount(), peer)
+}
+
+func matchRequest(rule *proto.Rule, req *authz.AttributeContext_Request) bool {
+	log.WithFields(log.Fields{
+		"request": req,
+	}).Debug("Matching request.")
+	return matchHTTP(rule.GetHttp(), req.GetHttp())
+}
+
+func matchServiceAccounts(saMatch *proto.ServiceAccountSelector, peer *authz.AttributeContext_Peer) bool {
+	principle := peer.GetPrincipal()
+	labels := peer.GetLabels()
+	log.WithFields(log.Fields{
+		"peer":   principle,
+		"labels": labels,
+		"rule":   saMatch},
+	).Debug("Matching service account.")
+	accountName, namespace, err := parseSpiffeId(principle)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"principle": principle,
+			"msg":       err,
+		}).Warn("Unable to parse authenticated principle as SPIFFE ID.")
+		return false
+	}
+	log.WithFields(log.Fields{
+		"name":      accountName,
+		"namespace": namespace,
+	}).Debug("Parsed SPIFFE ID.")
+	if saMatch == nil {
+		log.Debug("nil ServiceAccountMatch.  Return true.")
+		return true
+	}
+	return matchServiceAccountName(saMatch.GetNames(), accountName) &&
+		matchServiceAccountLabels(saMatch.GetLabelSelector(), labels)
+}
+
+// Parse an Istio SPIFFE ID and extract the service account name and namespace.
+func parseSpiffeId(id string) (name, namespace string, err error) {
+	// Init the regexp the first time this is called, and store it in the package namespace.
+	if spiffeIdRegExp == nil {
+		// We drop the returned error here, since we are compiling
+		spiffeIdRegExp, _ = regexp.Compile(SPIFFE_ID_PATTERN)
+	}
+	match := spiffeIdRegExp.FindStringSubmatch(id)
+	if match == nil {
+		err = fmt.Errorf("expected match %s, got %s", SPIFFE_ID_PATTERN, id)
+	} else {
+		name = match[2]
+		namespace = match[1]
+	}
+	return
+}
+
+func matchServiceAccountName(names []string, name string) bool {
+	log.WithFields(log.Fields{
+		"names": names,
+		"name":  name,
+	}).Debug("Matching service account name")
+	if len(names) == 0 {
+		log.Debug("No service account names on rule.")
+		return true
+	}
+	for _, name2 := range names {
```

Comment:
another tiny bit of Go best practice is, if a variable scope is small then its name should be shorter, and longer the scope more meaningful the name. So, I think you can just use `n` here instead of `name2`

Reply:
+1